File: ruby/lib/apparel_monster.rb
# frozen_string_literal: true

require "json"
require "net/http"
require "uri"

require_relative "apparel_monster/version"

# Apparel Monster SDK: a dependency-free client for the agent commerce API.
#
# No API key appears in this file and that is not an omission. search, product,
# cart, checkout, wallet, price-watch, batch and /ask are anonymous. Only order
# history (+track+) needs OAuth. If something asks you for an Apparel Monster
# API key, it is not us.
#
# Standard library only — net/http, no faraday, no httparty — so installing
# this pulls in nothing and it works inside an agent sandbox with no gem
# server.
#
#   store = ApparelMonster::Client.new
#   hit   = store.search("denim shirt", limit: 1)["products"].first
#   store.add_to_cart(hit["variants"].first["id"])
module ApparelMonster
  DEFAULT_BASE = "https://apparel.monster"
  API = "/api/v1"

  # Raised for any non-2xx response.
  #
  # The interesting attribute is +missing+: the API answers a bad call by
  # naming every field it needed and did not get, what else each one is
  # accepted as, and an example value. That detail is folded into the message
  # too, because the common case is a model reading +message+ and nothing else.
  class Error < StandardError
    attr_reader :status, :code, :missing, :next_call, :contract, :body, :url

    def initialize(status, body, url)
      @status = status
      @body = body || {}
      @code = @body["error"]
      @missing = @body["missing"]
      @next_call = @body["next_call"]
      @contract = @body["contract"]
      @url = url
      super(describe)
    end

    private

    def describe
      parts = [@body["message"] || "Request failed with #{@status}"]

      if @missing.is_a?(Array) && @missing.any?
        described = @missing.map do |field|
          names = (field["accepts"] || [field["field"]]).join(" or ")
          "#{field['field']} (send as #{names}, e.g. #{field['example'].inspect})"
        end
        parts << "Missing: #{described.join('; ')}"
      end

      parts << "Next call: #{JSON.generate(@next_call)}" if @next_call

      parts.join(" ")
    end
  end

  # A client for one shopping session.
  #
  # Two things this does that a raw Net::HTTP call does not: it carries the
  # session id the API mints on the first search (forgetting it is the usual
  # way an integration ends up with four empty carts and no order), and it
  # identifies itself so the store can attribute the resulting order to this
  # SDK rather than to "some script".
  class Client
    attr_accessor :session_id, :access_token
    attr_reader :base_url

    # @param base_url [String] override the origin
    # @param session_id [String, nil] resume an existing agent session
    # @param access_token [String, nil] OAuth token, only needed by #track
    # @param timeout [Numeric] seconds
    # @param cli [Boolean] identify as the CLI rather than the library
    def initialize(base_url: DEFAULT_BASE, session_id: nil, access_token: nil, timeout: 30, cli: false)
      @base_url = base_url.to_s.sub(%r{/+\z}, "")
      @session_id = session_id
      @access_token = access_token
      @timeout = timeout
      @client = cli ? "ruby-cli" : "ruby-sdk"
    end

    # ---- catalog -------------------------------------------------------

    # Search the catalog, or browse it when +query+ is nil.
    #
    # Mints the session id every later call needs, so this is almost always the
    # first call you make.
    def search(query = nil, category: nil, sort: nil, limit: nil, page: nil)
      post("/search", "query" => query, "category" => category, "sort" => sort, "limit" => limit, "page" => page)
    end

    # Full detail for product ids or variant ids, mixed freely.
    def product(ids)
      post("/product", "ids" => Array(ids).map(&:to_s))
    end

    # ---- cart ----------------------------------------------------------

    # Add a VARIANT — not a product. A product with five sizes has five
    # variants and five prices, and adding a product id is the most common
    # mistake made against this API.
    def add_to_cart(variant_id, quantity = 1)
      post("/cart", "action" => "add", "variant_id" => variant_id, "quantity" => quantity)
    end

    def update_cart_item(variant_id, quantity)
      post("/cart", "action" => "update", "variant_id" => variant_id, "quantity" => quantity)
    end

    def remove_from_cart(variant_id)
      post("/cart", "action" => "remove", "variant_id" => variant_id)
    end

    def view_cart
      post("/cart", "action" => "view")
    end

    # Set the email and address; returns the cart with +shipping_options+
    # priced against it.
    #
    # Required: email (once), line_one, city, postal_code. A missing one raises
    # ApparelMonster::Error naming all of them at once, with the aliases each
    # is accepted under.
    def set_billing(address = {})
      post("/cart", stringify(address).merge("action" => "billing"))
    end

    def set_shipping(shipping_id, address = {})
      post("/cart", stringify(address).merge("action" => "shipping", "shipping_id" => shipping_id))
    end

    def apply_coupon(code)
      post("/cart", "action" => "coupon", "coupon_code" => code)
    end

    # Attach a payment token. The field is +token+ — not +payment_token+, not
    # +card_token+. Demo store on a test gateway: any non-empty string is
    # accepted and nothing is ever charged.
    def set_payment(token)
      post("/cart", "action" => "payment", "token" => token)
    end

    # ---- checkout ------------------------------------------------------

    # Place the order. Idempotent per session.
    def checkout(callback_url: nil)
      post("/checkout", "callback_url" => callback_url)
    end

    # A hosted Apple Pay / Google Pay link, for when a human can tap.
    #
    # The shortcut past the whole cart flow: hand it variant ids and it returns
    # a URL that collects the shopper's email, address and card on their own
    # device, so your agent never handles any of them.
    def wallet(items)
      normalised = Array(items).map do |item|
        item.is_a?(Hash) ? stringify(item) : { "variant_id" => item.to_s, "quantity" => 1 }
      end
      post("/wallet", "items" => normalised)
    end

    # Order status and shipments. The only call that needs OAuth.
    def track(params = {})
      post("/track", stringify(params))
    end

    # ---- price watch ---------------------------------------------------

    # Be told when a variant reaches a target price. Answers 202 with a
    # +poll_url+ and a +job_id+; a target at or above today's price is not
    # queued at all and comes back +already_met+.
    def watch_price(variant_id, target_price, callback_url: nil)
      post("/price-watch",
           "variant_id" => variant_id, "target_price" => target_price, "callback_url" => callback_url)
    end

    def price_watch(watch_id)
      request(Net::HTTP::Get, "#{@base_url}#{API}/price-watch/#{URI.encode_www_form_component(watch_id)}")
    end

    # ---- batch ---------------------------------------------------------

    # Up to 10 operations in one round trip, executed in order. Each result
    # carries its own status: the batch answers 200 whenever it ran, even if
    # everything inside it failed.
    def batch(operations, stop_on_error: false)
      raise ArgumentError, "batch takes at most 10 operations" if operations.length > 10

      post("/batch", "operations" => operations, "stop_on_error" => stop_on_error)
    end

    # ---- natural language ----------------------------------------------

    # Ask in plain language (NLWeb); returns schema.org Product items. There is
    # no model behind it — it runs the same catalog search — which is why it
    # returns nothing rather than inventing a product.
    def ask(query, limit: nil)
      params = { "query" => query }
      params["limit"] = limit if limit
      request(Net::HTTP::Get, "#{@base_url}/ask?#{URI.encode_www_form(params)}")
    end

    # The same question, streamed: yields [event, data] as +start+, one
    # +result+ per hit, then +complete+.
    def ask_stream(query, limit: nil)
      return enum_for(:ask_stream, query, limit: limit) unless block_given?

      params = { "query" => query, "streaming" => "true" }
      params["limit"] = limit if limit
      uri = URI("#{@base_url}/ask?#{URI.encode_www_form(params)}")

      http(uri).request(get_request(uri, "text/event-stream")) do |response|
        event = "message"
        data = +""
        response.read_body do |chunk|
          chunk.each_line do |raw|
            line = raw.chomp
            if line.start_with?("event:")
              event = line[6..].strip
            elsif line.start_with?("data:")
              data << line[5..].strip
            elsif line.empty?
              # A blank line terminates an SSE frame; anything before one is a
              # partial frame and must not be yielded.
              unless data.empty?
                begin
                  yield [event, JSON.parse(data)]
                rescue JSON::ParserError
                  nil
                end
              end
              event = "message"
              data = +""
            end
          end
        end
      end
    end

    # ---- plumbing ------------------------------------------------------

    def headers(accept = "application/json")
      suffix = @client == "ruby-cli" ? "-cli" : ""
      result = {
        "Content-Type" => "application/json",
        "Accept" => accept,
        # Two spellings of one fact: the User-Agent is what the edge sees
        # without the application being involved, X-Agent-Client is what
        # survives an environment that rewrites User-Agent.
        "User-Agent" => "apparel-monster-ruby#{suffix}/#{VERSION} (+https://github.com/mluggy/apparel-monster-dev)",
        "X-Agent-Client" => "#{@client}/#{VERSION}"
      }
      result["Authorization"] = "Bearer #{@access_token}" if @access_token
      result
    end

    private

    def stringify(hash)
      (hash || {}).each_with_object({}) { |(k, v), out| out[k.to_s] = v }
    end

    def with_session(body)
      return body unless @session_id

      { "session" => { "id" => @session_id } }.merge(body)
    end

    def post(path, body = {})
      payload = with_session(body.reject { |_k, v| v.nil? })
      request(Net::HTTP::Post, "#{@base_url}#{API}#{path}", JSON.generate(payload))
    end

    def http(uri)
      client = Net::HTTP.new(uri.host, uri.port)
      client.use_ssl = uri.scheme == "https"
      client.open_timeout = @timeout
      client.read_timeout = @timeout
      client
    end

    def get_request(uri, accept)
      Net::HTTP::Get.new(uri).tap { |r| headers(accept).each { |k, v| r[k] = v } }
    end

    def request(verb, url, body = nil)
      uri = URI(url)
      req = verb.new(uri)
      headers.each { |k, v| req[k] = v }
      req.body = body if body

      response = http(uri).request(req)
      parsed = begin
        JSON.parse(response.body.to_s)
      rescue JSON::ParserError
        nil
      end

      raise Error.new(response.code.to_i, parsed, url) unless response.is_a?(Net::HTTPSuccess)

      # The session arrives on the first search and is reused from then on.
      if parsed.is_a?(Hash) && parsed.dig("session", "id")
        @session_id = parsed["session"]["id"]
      end

      parsed
    end
  end
end

File: go/apparelmonster.go
// Package apparelmonster is a dependency-free client for the Apparel Monster
// agent commerce API.
//
// No API key appears in this package and that is not an omission. Search,
// Product, Cart, Checkout, Wallet, PriceWatch, Batch and Ask are anonymous.
// Only order history (Track) needs OAuth. If something asks you for an Apparel
// Monster API key, it is not us.
//
// Standard library only, so adding this to a project pulls in nothing.
//
//	store := apparelmonster.New()
//	found, err := store.Search(ctx, "denim shirt", apparelmonster.SearchOptions{Limit: 1})
//	if err != nil {
//	    log.Fatal(err)
//	}
//	_, err = store.AddToCart(ctx, found.Products[0].Variants[0].ID, 1)
//
// apparel.monster is a demonstration storefront: orders are placed against a
// test gateway, nothing is fulfilled and no money moves.
package apparelmonster

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

// Version is reported in the User-Agent and X-Agent-Client headers.
const Version = "1.0.2"

const (
	defaultBaseURL = "https://apparel.monster"
	apiPrefix      = "/api/v1"
)

// MissingField describes one field the API needed and did not get.
//
// Accepts lists every name the field is taken under: this API answers to
// line_one or address1, postal_code or zipcode or zip, because agents arrive
// carrying schemas from other stores.
type MissingField struct {
	Field       string   `json:"field"`
	Accepts     []string `json:"accepts"`
	Description string   `json:"description"`
	Example     any      `json:"example"`
}

// Error is returned for any non-2xx response.
//
// Missing is the interesting part: rather than a generic 4xx, the API names
// every field it needed, and Error's message repeats them, because the common
// case is a caller logging err and nothing else.
type Error struct {
	Status   int
	Code     string
	Message  string
	Missing  []MissingField
	NextCall map[string]any
	Contract map[string]any
	URL      string
}

func (e *Error) Error() string {
	parts := []string{e.Message}
	if len(parts[0]) == 0 {
		parts[0] = fmt.Sprintf("request failed with %d", e.Status)
	}

	if len(e.Missing) > 0 {
		described := make([]string, 0, len(e.Missing))
		for _, field := range e.Missing {
			names := field.Accepts
			if len(names) == 0 {
				names = []string{field.Field}
			}
			described = append(described, fmt.Sprintf("%s (send as %s, e.g. %v)",
				field.Field, strings.Join(names, " or "), field.Example))
		}
		parts = append(parts, "Missing: "+strings.Join(described, "; "))
	}

	if e.NextCall != nil {
		if encoded, err := json.Marshal(e.NextCall); err == nil {
			parts = append(parts, "Next call: "+string(encoded))
		}
	}

	return strings.Join(parts, " ")
}

// Money mirrors the API's currency object.
type Money struct {
	ID string `json:"id"`
}

// Variant is a purchasable item. Prices live here, never on the Product: a
// product with five sizes has five variants and five prices.
type Variant struct {
	ID             string            `json:"id"`
	SKU            string            `json:"sku"`
	Title          string            `json:"title"`
	Price          float64           `json:"price"`
	EffectivePrice float64           `json:"effective_price"`
	Currency       Money             `json:"currency"`
	Available      bool              `json:"available"`
	Options        map[string]string `json:"options"`
	ImageURL       string            `json:"image_url"`
	URL            string            `json:"url"`
}

// Image is one product image.
type Image struct {
	URL string `json:"url"`
}

// Product is a catalog entry with its purchasable variants.
type Product struct {
	ID       string    `json:"id"`
	Title    string    `json:"title"`
	URL      string    `json:"url"`
	Category string    `json:"category"`
	Images   []Image   `json:"images"`
	Variants []Variant `json:"variants"`
}

// Session identifies one shopping session; the client carries it for you.
type Session struct {
	ID string `json:"id"`
}

// Totals is the money breakdown. Read Total; do not re-derive it.
type Totals struct {
	Subtotal float64 `json:"subtotal"`
	Discount float64 `json:"discount"`
	Tax      float64 `json:"tax"`
	Shipping float64 `json:"shipping"`
	Total    float64 `json:"total"`
}

// ShippingOption is one priced delivery choice.
type ShippingOption struct {
	ID         string  `json:"id"`
	Name       string  `json:"name"`
	Price      float64 `json:"price"`
	Delivery   string  `json:"delivery"`
	IsSelected bool    `json:"is_selected"`
}

// SearchResponse is the result of a catalog search.
type SearchResponse struct {
	Session  Session   `json:"session"`
	Products []Product `json:"products"`
	Page     int       `json:"page"`
	Limit    int       `json:"limit"`
	Total    int       `json:"total"`
	HasMore  bool      `json:"has_more"`
}

// CartResponse is the cart as it stands.
type CartResponse struct {
	Session         Session          `json:"session"`
	State           string           `json:"state"`
	Currency        Money            `json:"currency"`
	Products        []Product        `json:"products"`
	Totals          Totals           `json:"totals"`
	ShippingOptions []ShippingOption `json:"shipping_options"`
}

// Address is the shipping and billing address.
//
// Required: LineOne, City, PostalCode, and Email once per session. Country
// defaults to US. A missing field comes back as an *Error naming all of them.
type Address struct {
	Email      string `json:"email,omitempty"`
	FirstName  string `json:"first_name,omitempty"`
	LastName   string `json:"last_name,omitempty"`
	LineOne    string `json:"line_one,omitempty"`
	LineTwo    string `json:"line_two,omitempty"`
	City       string `json:"city,omitempty"`
	RegionID   string `json:"region_id,omitempty"`
	PostalCode string `json:"postal_code,omitempty"`
	CountryID  string `json:"country_id,omitempty"`
	Phone      string `json:"phone,omitempty"`
}

// AskResult is one NLWeb hit.
type AskResult struct {
	URL          string         `json:"url"`
	Name         string         `json:"name"`
	Site         string         `json:"site"`
	Score        float64        `json:"score"`
	Description  string         `json:"description"`
	SchemaObject map[string]any `json:"schema_object"`
}

// AskResponse is an NLWeb result list.
type AskResponse struct {
	Meta struct {
		ResponseType string `json:"response_type"`
		Version      string `json:"version"`
		Query        string `json:"query"`
		Site         string `json:"site"`
		Count        int    `json:"count"`
	} `json:"_meta"`
	Results []AskResult `json:"results"`
}

// BatchOperation is one request inside a batch.
type BatchOperation struct {
	ID     string         `json:"id,omitempty"`
	Method string         `json:"method,omitempty"`
	Path   string         `json:"path"`
	Body   map[string]any `json:"body,omitempty"`
}

// BatchResponse carries one result per operation, each with its own status.
type BatchResponse struct {
	Count        int  `json:"count"`
	Requested    int  `json:"requested"`
	Succeeded    int  `json:"succeeded"`
	StoppedEarly bool `json:"stopped_early"`
	Results      []struct {
		ID     string          `json:"id"`
		Status int             `json:"status"`
		Body   json.RawMessage `json:"body"`
	} `json:"results"`
}

// StreamEvent is one Server-Sent Event from Ask streaming.
type StreamEvent struct {
	Event string
	Data  json.RawMessage
}

// Client talks to one store, carrying one shopping session.
//
// Two things it does that a raw http.Client does not: it carries the session
// id the API mints on the first search (forgetting it is the usual way an
// integration ends up with four empty carts and no order), and it identifies
// itself so the store can attribute the resulting order to this SDK.
//
// A Client is safe for sequential use; the session it carries makes concurrent
// use across goroutines meaningless rather than merely unsafe. Use one Client
// per shopper.
type Client struct {
	BaseURL     string
	SessionID   string
	AccessToken string
	HTTPClient  *http.Client

	client string
}

// Option configures a Client.
type Option func(*Client)

// WithBaseURL points the client at another origin.
func WithBaseURL(base string) Option {
	return func(c *Client) { c.BaseURL = strings.TrimRight(base, "/") }
}

// WithSession resumes an existing agent session.
func WithSession(id string) Option {
	return func(c *Client) { c.SessionID = id }
}

// WithAccessToken supplies the OAuth token that Track needs.
func WithAccessToken(token string) Option {
	return func(c *Client) { c.AccessToken = token }
}

// WithHTTPClient injects an http.Client, for timeouts or tests.
func WithHTTPClient(client *http.Client) Option {
	return func(c *Client) { c.HTTPClient = client }
}

// AsCLI identifies this client as the command line tool rather than the
// library, so the store can tell a person poking at the API from an
// integration running in production.
func AsCLI() Option {
	return func(c *Client) { c.client = "go-cli" }
}

// New builds a client against the public store.
func New(options ...Option) *Client {
	client := &Client{
		BaseURL:    defaultBaseURL,
		HTTPClient: &http.Client{Timeout: 30 * time.Second},
		client:     "go-sdk",
	}
	for _, option := range options {
		option(client)
	}
	return client
}

// SearchOptions narrows a catalog search.
type SearchOptions struct {
	Category string
	Sort     string
	Limit    int
	Page     int
}

// Search searches the catalog, or browses it when query is empty.
//
// It mints the session id every later call needs, so it is almost always the
// first call you make.
func (c *Client) Search(ctx context.Context, query string, options SearchOptions) (*SearchResponse, error) {
	body := map[string]any{}
	if query != "" {
		body["query"] = query
	}
	if options.Category != "" {
		body["category"] = options.Category
	}
	if options.Sort != "" {
		body["sort"] = options.Sort
	}
	if options.Limit > 0 {
		body["limit"] = options.Limit
	}
	if options.Page > 0 {
		body["page"] = options.Page
	}

	var out SearchResponse
	err := c.post(ctx, "/search", body, &out)
	return &out, err
}

// Product returns full detail for product ids or variant ids, mixed freely.
func (c *Client) Product(ctx context.Context, ids ...string) (*SearchResponse, error) {
	var out SearchResponse
	err := c.post(ctx, "/product", map[string]any{"ids": ids}, &out)
	return &out, err
}

// AddToCart adds a VARIANT — not a product. Adding a product id is the most
// common mistake made against this API.
func (c *Client) AddToCart(ctx context.Context, variantID string, quantity int) (*CartResponse, error) {
	if quantity <= 0 {
		quantity = 1
	}
	return c.cart(ctx, map[string]any{"action": "add", "variant_id": variantID, "quantity": quantity})
}

// UpdateCartItem changes the quantity of a line already in the cart.
func (c *Client) UpdateCartItem(ctx context.Context, variantID string, quantity int) (*CartResponse, error) {
	return c.cart(ctx, map[string]any{"action": "update", "variant_id": variantID, "quantity": quantity})
}

// RemoveFromCart removes a line entirely.
func (c *Client) RemoveFromCart(ctx context.Context, variantID string) (*CartResponse, error) {
	return c.cart(ctx, map[string]any{"action": "remove", "variant_id": variantID})
}

// ViewCart returns the cart with its totals.
func (c *Client) ViewCart(ctx context.Context) (*CartResponse, error) {
	return c.cart(ctx, map[string]any{"action": "view"})
}

// SetBilling sets the email and address, and returns the cart with
// ShippingOptions priced against it. This call is what makes them exist.
func (c *Client) SetBilling(ctx context.Context, address Address) (*CartResponse, error) {
	body, err := toMap(address)
	if err != nil {
		return nil, err
	}
	body["action"] = "billing"
	return c.cart(ctx, body)
}

// SetShipping chooses a shipping option by id.
func (c *Client) SetShipping(ctx context.Context, shippingID string) (*CartResponse, error) {
	return c.cart(ctx, map[string]any{"action": "shipping", "shipping_id": shippingID})
}

// ApplyCoupon applies a promotion code.
func (c *Client) ApplyCoupon(ctx context.Context, code string) (*CartResponse, error) {
	return c.cart(ctx, map[string]any{"action": "coupon", "coupon_code": code})
}

// SetPayment attaches a payment token.
//
// The field is token — not payment_token, not card_token. This is a demo store
// on a test gateway: any non-empty string is accepted and nothing is charged.
func (c *Client) SetPayment(ctx context.Context, token string) (*CartResponse, error) {
	return c.cart(ctx, map[string]any{"action": "payment", "token": token})
}

// Checkout places the order. Idempotent per session.
func (c *Client) Checkout(ctx context.Context, callbackURL string) (map[string]any, error) {
	body := map[string]any{}
	if callbackURL != "" {
		body["callback_url"] = callbackURL
	}
	var out map[string]any
	err := c.post(ctx, "/checkout", body, &out)
	return out, err
}

// Wallet mints a hosted Apple Pay / Google Pay link for the given variants.
//
// The shortcut past the whole cart flow: the returned URL collects the
// shopper's email, address and card on their own device, so your agent never
// handles any of them.
func (c *Client) Wallet(ctx context.Context, variantIDs ...string) (map[string]any, error) {
	items := make([]map[string]any, 0, len(variantIDs))
	for _, id := range variantIDs {
		items = append(items, map[string]any{"variant_id": id, "quantity": 1})
	}
	var out map[string]any
	err := c.post(ctx, "/wallet", map[string]any{"items": items}, &out)
	return out, err
}

// Track returns order status and shipments. The only call needing OAuth: pass
// WithAccessToken.
func (c *Client) Track(ctx context.Context, params map[string]any) (map[string]any, error) {
	var out map[string]any
	err := c.post(ctx, "/track", params, &out)
	return out, err
}

// WatchPrice asks to be told when a variant reaches a target price.
//
// Answers 202 with a poll_url and a job_id. A target at or above today's price
// is not queued at all: it comes back already_met, so act on it now.
func (c *Client) WatchPrice(ctx context.Context, variantID string, target float64, callbackURL string) (map[string]any, error) {
	body := map[string]any{"variant_id": variantID, "target_price": target}
	if callbackURL != "" {
		body["callback_url"] = callbackURL
	}
	var out map[string]any
	err := c.post(ctx, "/price-watch", body, &out)
	return out, err
}

// PriceWatch polls a price watch by id.
func (c *Client) PriceWatch(ctx context.Context, id string) (map[string]any, error) {
	var out map[string]any
	err := c.get(ctx, c.BaseURL+apiPrefix+"/price-watch/"+url.PathEscape(id), "application/json", &out)
	return out, err
}

// Batch runs up to ten operations in one round trip, in order.
//
// Each result carries its own status: the batch answers 200 whenever it ran,
// even if every operation inside it failed.
func (c *Client) Batch(ctx context.Context, operations []BatchOperation, stopOnError bool) (*BatchResponse, error) {
	if len(operations) > 10 {
		return nil, fmt.Errorf("apparelmonster: batch takes at most 10 operations, got %d", len(operations))
	}
	var out BatchResponse
	err := c.post(ctx, "/batch", map[string]any{"operations": operations, "stop_on_error": stopOnError}, &out)
	return &out, err
}

// Ask queries in plain language (NLWeb) and returns schema.org Products.
//
// There is no model behind it — it runs the same catalog search — which is why
// it returns nothing rather than inventing a product.
func (c *Client) Ask(ctx context.Context, query string, limit int) (*AskResponse, error) {
	params := url.Values{"query": {query}}
	if limit > 0 {
		params.Set("limit", strconv.Itoa(limit))
	}
	var out AskResponse
	err := c.get(ctx, c.BaseURL+"/ask?"+params.Encode(), "application/json", &out)
	return &out, err
}

// AskStream is the same question, streamed: start, one result per hit, then
// complete. The channel is closed when the stream ends.
func (c *Client) AskStream(ctx context.Context, query string, limit int) (<-chan StreamEvent, error) {
	params := url.Values{"query": {query}, "streaming": {"true"}}
	if limit > 0 {
		params.Set("limit", strconv.Itoa(limit))
	}

	request, err := c.newRequest(ctx, http.MethodGet, c.BaseURL+"/ask?"+params.Encode(), nil, "text/event-stream")
	if err != nil {
		return nil, err
	}

	response, err := c.HTTPClient.Do(request)
	if err != nil {
		return nil, err
	}
	if response.StatusCode != http.StatusOK {
		defer response.Body.Close()
		return nil, c.asError(response, c.BaseURL+"/ask")
	}

	events := make(chan StreamEvent)
	go func() {
		defer close(events)
		defer response.Body.Close()

		scanner := bufio.NewScanner(response.Body)
		scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)

		event := "message"
		var data strings.Builder

		for scanner.Scan() {
			line := scanner.Text()
			switch {
			case strings.HasPrefix(line, "event:"):
				event = strings.TrimSpace(line[6:])
			case strings.HasPrefix(line, "data:"):
				data.WriteString(strings.TrimSpace(line[5:]))
			case line == "":
				// A blank line terminates an SSE frame; anything before one is
				// a partial frame and must not be emitted.
				if data.Len() > 0 {
					select {
					case events <- StreamEvent{Event: event, Data: json.RawMessage(data.String())}:
					case <-ctx.Done():
						return
					}
				}
				event = "message"
				data.Reset()
			}
		}
	}()

	return events, nil
}

// ---- plumbing --------------------------------------------------------------

func (c *Client) cart(ctx context.Context, body map[string]any) (*CartResponse, error) {
	var out CartResponse
	err := c.post(ctx, "/cart", body, &out)
	return &out, err
}

func (c *Client) userAgent() string {
	suffix := ""
	if c.client == "go-cli" {
		suffix = "-cli"
	}
	return fmt.Sprintf("apparel-monster-go%s/%s (+https://github.com/mluggy/apparel-monster-dev)", suffix, Version)
}

func (c *Client) newRequest(ctx context.Context, method, url string, body io.Reader, accept string) (*http.Request, error) {
	request, err := http.NewRequestWithContext(ctx, method, url, body)
	if err != nil {
		return nil, err
	}

	request.Header.Set("Content-Type", "application/json")
	request.Header.Set("Accept", accept)
	// Two spellings of one fact: the User-Agent is what the edge sees without
	// the application being involved, X-Agent-Client is what survives an
	// environment that rewrites User-Agent.
	request.Header.Set("User-Agent", c.userAgent())
	request.Header.Set("X-Agent-Client", c.client+"/"+Version)
	if c.AccessToken != "" {
		request.Header.Set("Authorization", "Bearer "+c.AccessToken)
	}

	return request, nil
}

func (c *Client) post(ctx context.Context, path string, body map[string]any, out any) error {
	if body == nil {
		body = map[string]any{}
	}
	if c.SessionID != "" {
		if _, taken := body["session"]; !taken {
			body["session"] = map[string]any{"id": c.SessionID}
		}
	}

	encoded, err := json.Marshal(body)
	if err != nil {
		return err
	}

	full := c.BaseURL + apiPrefix + path
	request, err := c.newRequest(ctx, http.MethodPost, full, bytes.NewReader(encoded), "application/json")
	if err != nil {
		return err
	}

	return c.do(request, full, out)
}

func (c *Client) get(ctx context.Context, url, accept string, out any) error {
	request, err := c.newRequest(ctx, http.MethodGet, url, nil, accept)
	if err != nil {
		return err
	}
	return c.do(request, url, out)
}

func (c *Client) do(request *http.Request, url string, out any) error {
	response, err := c.HTTPClient.Do(request)
	if err != nil {
		return err
	}
	defer response.Body.Close()

	payload, err := io.ReadAll(response.Body)
	if err != nil {
		return err
	}

	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return errorFromBody(response.StatusCode, payload, url)
	}

	if out != nil && len(payload) > 0 {
		if err := json.Unmarshal(payload, out); err != nil {
			return fmt.Errorf("apparelmonster: could not decode response from %s: %w", url, err)
		}
	}

	// The session arrives on the first search and is reused from then on.
	var envelope struct {
		Session Session `json:"session"`
	}
	if err := json.Unmarshal(payload, &envelope); err == nil && envelope.Session.ID != "" {
		c.SessionID = envelope.Session.ID
	}

	return nil
}

func (c *Client) asError(response *http.Response, url string) error {
	payload, _ := io.ReadAll(response.Body)
	return errorFromBody(response.StatusCode, payload, url)
}

func errorFromBody(status int, payload []byte, url string) error {
	var body struct {
		Error    string         `json:"error"`
		Message  string         `json:"message"`
		Missing  []MissingField `json:"missing"`
		NextCall map[string]any `json:"next_call"`
		Contract map[string]any `json:"contract"`
	}
	_ = json.Unmarshal(payload, &body)

	return &Error{
		Status:   status,
		Code:     body.Error,
		Message:  body.Message,
		Missing:  body.Missing,
		NextCall: body.NextCall,
		Contract: body.Contract,
		URL:      url,
	}
}

// toMap round-trips through JSON so the struct tags stay the single definition
// of what each field is called on the wire.
func toMap(value any) (map[string]any, error) {
	encoded, err := json.Marshal(value)
	if err != nil {
		return nil, err
	}
	out := map[string]any{}
	if err := json.Unmarshal(encoded, &out); err != nil {
		return nil, err
	}
	return out, nil
}
